src/Avalonia.Visuals/Rendering/SceneGraph/SceneBuilder.cs

```diff
@@ -0,0 +1,389 @@
+                    if (node != null)
+                    {
+                        // The control has been hidden so remove it from its parent and deindex the
+                        // node and its descendents.
+                        ((VisualNode)node.Parent)?.RemoveChild(node);
+                        Deindex(scene, node);
+                        return true;
+                    }
+                }
+            }
+            else if (node != null)
+            {
+                // The control has been removed so remove it from its parent and deindex the
+                // node and its descendents.
+                var trim = FindFirstDeadAncestor(scene, node);
+                ((VisualNode)trim.Parent).RemoveChild(trim);
+                Deindex(scene, trim);
+                return true;
+            }
+
+            return false;
+        }
+
+        private static VisualNode FindExistingAncestor(Scene scene, IVisual visual)
+        {
+            var node = scene.FindNode(visual);
+
+            while (node == null && visual.IsVisible)
+            {
+                visual = visual.VisualParent;
+                node = scene.FindNode(visual);
+            }
+
+            return visual.IsVisible ? (VisualNode)node : null;
+        }
+
+        private static VisualNode FindFirstDeadAncestor(Scene scene, IVisualNode node)
+        {
+            var parent = node.Parent;
+
+            while (parent.Visual.VisualRoot == null)
+            {
+                node = parent;
+                parent = node.Parent;
+            }
+
+            return (VisualNode)node;
+        }
+
+        private static void Update(DrawingContext context, Scene scene, VisualNode node, Rect clip, bool forceRecurse)
+        {
+            var visual = node.Visual;
+            var opacity = visual.Opacity;
+            var clipToBounds = visual.ClipToBounds;
+            var bounds = new Rect(visual.Bounds.Size);
+            var contextImpl = (DeferredDrawingContextImpl)context.PlatformImpl;
+
+            contextImpl.Layers.Find(node.LayerRoot)?.Dirty.Add(node.Bounds);
+
+            if (visual.IsVisible)
+            {
+                var m = Matrix.CreateTranslation(visual.Bounds.Position);
+
+                var renderTransform = Matrix.Identity;
+
+                if (visual.RenderTransform != null)
+                {
+                    var origin = visual.RenderTransformOrigin.ToPixels(new Size(visual.Bounds.Width, visual.Bounds.Height));
+                    var offset = Matrix.CreateTranslation(origin);
+                    renderTransform = (-offset) * visual.RenderTransform.Value * (offset);
+                }
+
+                m = renderTransform * m;
+
+                using (contextImpl.BeginUpdate(node))
+                using (context.PushPostTransform(m))
+                using (context.PushTransformContainer())
+                {
+                    var startLayer = opacity < 1 || visual.OpacityMask != null;
```

Comment:
Unfortunately doing this at the moment causes RenderTest to run at 1 FPS on D2D. I think we should do this after we fix the use of WIC bitmaps for layers in D2D.